In [199]:
#Se cargan las librerias

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [200]:
# Cargar el dataset
data = pd.read_csv('data/heart-disease.csv')

# Exploración de datos
print("Información del dataset:")
print(data.info())

Información del dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB
None


In [201]:
print("\nEstadísticas descriptivas:")
print(data.describe())


Estadísticas descriptivas:
              age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   

In [202]:
print("\nValores únicos en la variable 'target':")
print(data['target'].unique())


Valores únicos en la variable 'target':
[1 0]


Ahora se usaran distintos algoritmos de Machine Learning, para poder aumentar la precisión del modelo anteriormente mostrado.

## Regresión logística:

In [203]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# Dividir el dataset en características (X) y variable objetivo (y)
X = data.drop('target', axis=1)
y = data['target']

# Dividir el dataset en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Crear el modelo de regresión logística
model = LogisticRegression(max_iter=1000)

# Entrenar el modelo
model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión de la regresión logística:", accuracy)


Precisión de la regresión logística: 0.8852459016393442


## Árbol de decisión:

In [204]:
from sklearn.tree import DecisionTreeClassifier

# Crear el modelo de árbol de decisión
model = DecisionTreeClassifier()

# Entrenar el modelo
model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión del árbol de decisión:", accuracy)


Precisión del árbol de decisión: 0.8360655737704918


## Bosque aleatorio (Random Forest):

In [205]:
from sklearn.ensemble import RandomForestClassifier

# Crear el modelo de bosque aleatorio
model = RandomForestClassifier()

# Entrenar el modelo
model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión del bosque aleatorio:", accuracy)


Precisión del bosque aleatorio: 0.8360655737704918


## Máquinas de vectores de soporte (Support Vector Machines, SVM)

In [206]:
from sklearn.svm import SVC

# Crear el modelo SVM
model = SVC()

# Entrenar el modelo
model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión de las Máquinas de Vectores de Soporte:", accuracy)


Precisión de las Máquinas de Vectores de Soporte: 0.7049180327868853


Aquí está la interpretación de los resultados de cada algoritmo:

Regresión logística: La precisión de 0.8852 indica que el 88.52% de las predicciones realizadas por el modelo de regresión logística fueron correctas.

Árbol de decisión: La precisión de 0.8033 significa que el 80.33% de las predicciones realizadas por el modelo de árbol de decisión fueron correctas.

Bosque aleatorio: La precisión de 0.8689 indica que el 86.89% de las predicciones realizadas por el modelo de bosque aleatorio fueron correctas.

Máquinas de Vectores de Soporte (SVM): La precisión de 0.7049 significa que el 70.49% de las predicciones realizadas por el modelo de Máquinas de Vectores de Soporte fueron correctas.

## Validación cruzada:

In [207]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
import numpy as np

# Dividir el dataset en características (X) y variable objetivo (y)
X = data.drop('target', axis=1)
y = data['target']

# Inicializar los modelos
logistic_regression = LogisticRegression(max_iter=100000)  # Aumenta el número de iteraciones
decision_tree = DecisionTreeClassifier()
random_forest = RandomForestClassifier()
svm = SVC()

# Realizar la validación cruzada para cada modelo
models = [logistic_regression, decision_tree, random_forest, svm]
model_names = ['Regresión logística', 'Árbol de decisión', 'Bosque aleatorio', 'Máquinas de Vectores de Soporte']

for model, name in zip(models, model_names):
    scores = cross_val_score(model, X, y, cv=5)  # Realizar validación cruzada con 5 subdivisiones
    print(f"Precisión promedio de {name}: {np.mean(scores)}")


Precisión promedio de Regresión logística: 0.8282513661202187
Precisión promedio de Árbol de decisión: 0.7687978142076503
Precisión promedio de Bosque aleatorio: 0.8249726775956283
Precisión promedio de Máquinas de Vectores de Soporte: 0.6434972677595628


Ahora se realizara el ajuste de hiperparámetros para obtener una mejor precisión

## Regresión logística

In [208]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

# Crear el modelo de regresión logística
model = LogisticRegression(solver='lbfgs', max_iter=1000000)

# Definir los hiperparámetros que deseas ajustar
param_grid = {'C': [0.1, 1.0, 10.0], 'penalty': ['l2']}  # Usar 'l2' en lugar de 'l1'

# Realizar la búsqueda de hiperparámetros utilizando Grid Search
grid_search = GridSearchCV(model, param_grid, cv=5)
grid_search.fit(X_train, y_train)

# Obtener el mejor modelo y los mejores hiperparámetros
best_model = grid_search.best_estimator_
best_params = grid_search.best_params_

# Entrenar el mejor modelo con los mejores hiperparámetros
best_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = best_model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión de la regresión logística:", accuracy)



Precisión de la regresión logística: 0.8524590163934426


Se crea una cuadrícula de hiperparámetros param_grid que incluye diferentes valores para el parámetro de regularización C y el tipo de penalización penalty. Grid Search (GridSearchCV) se utiliza para buscar exhaustivamente los mejores hiperparámetros dentro de la cuadrícula definida. Después de la búsqueda, se obtiene el mejor modelo y los mejores hiperparámetros utilizando best_estimator_ y best_params_, respectivamente.

## Árbol de decisión

In [209]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.tree import DecisionTreeClassifier

# Crear el modelo de árbol de decisión
model = DecisionTreeClassifier()

# Definir los hiperparámetros que deseas ajustar
param_grid = {'max_depth': [None, 5, 10], 'min_samples_split': [2, 5, 10]}

# Realizar la búsqueda de hiperparámetros utilizando Randomized Search
randomized_search = RandomizedSearchCV(model, param_grid, cv=5, n_iter=9)  # Ajustar el valor de n_iter
randomized_search.fit(X_train, y_train)

# Obtener el mejor modelo y los mejores hiperparámetros
best_model = randomized_search.best_estimator_
best_params = randomized_search.best_params_

# Entrenar el mejor modelo con los mejores hiperparámetros
best_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = best_model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión del árbol de decisión:", accuracy)



Precisión del árbol de decisión: 0.7377049180327869


Se utiliza Randomized Search (RandomizedSearchCV) para buscar los mejores hiperparámetros para el modelo de árbol de decisión. Se define una cuadrícula de hiperparámetros param_grid que incluye diferentes valores para la profundidad máxima del árbol (max_depth) y el número mínimo de muestras requeridas para dividir un nodo interno (min_samples_split).

Y al ajustar n_iter=9, se asegura de que la búsqueda aleatoria se ejecute correctamente sin generar el mensaje de advertencia.

## Bosque aleatorio

In [210]:
from sklearn.model_selection import RandomizedSearchCV

# Crear el modelo de bosque aleatorio
model = RandomForestClassifier()

# Definir los hiperparámetros que deseas ajustar
param_grid = {'n_estimators': [100, 200, 500], 'max_depth': [None, 5, 10]}

# Realizar la búsqueda de hiperparámetros utilizando Randomized Search
randomized_search = RandomizedSearchCV(model, param_grid, cv=5 , n_iter=9)
randomized_search.fit(X_train, y_train)

# Obtener el mejor modelo y los mejores hiperparámetros
best_model = randomized_search.best_estimator_
best_params = randomized_search.best_params_

# Entrenar el mejor modelo con los mejores hiperparámetros
best_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = best_model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión del bosque aleatorio:", accuracy)


Precisión del bosque aleatorio: 0.8360655737704918


Se utiliza Randomized Search para buscar los mejores hiperparámetros para el modelo de bosque aleatorio. Se define una cuadrícula de hiperparámetros param_grid que incluye diferentes valores para el número de árboles en el bosque (n_estimators) y la profundidad máxima de los árboles (max_depth).

Y al ajustar n_iter=9, se asegura de que la búsqueda aleatoria se ejecute correctamente sin generar el mensaje de advertencia.

## Máquinas de Vectores de Soporte (SVM)

In [220]:
from sklearn.model_selection import GridSearchCV

# Crear el modelo SVM
model = SVC()

# Definir los hiperparámetros que deseas ajustar
param_grid = {'C': [0.1, 1.0, 10.0], 'kernel': ['linear', 'rbf']}

# Realizar la búsqueda de hiperparámetros utilizando Grid Search
grid_search = GridSearchCV(model, param_grid, cv=5)
grid_search.fit(X_train, y_train)

# Obtener el mejor modelo y los mejores hiperparámetros
best_model = grid_search.best_estimator_
best_params = grid_search.best_params_

# Entrenar el mejor modelo con los mejores hiperparámetros
best_model.fit(X_train, y_train)

# Realizar predicciones en el conjunto de prueba
y_pred = best_model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred)
print("Precisión de las Máquinas de Vectores de Soporte:", accuracy)


Precisión de las Máquinas de Vectores de Soporte: 0.8688524590163934


Se utiliza Grid Search para buscar los mejores hiperparámetros para el modelo SVM. Se define una cuadrícula de hiperparámetros param_grid que incluye diferentes valores para el parámetro de regularización C y el tipo de kernel (linear y rbf).